doc/developer/design/20230519_statement_logging.md

```diff
@@ -0,0 +1,165 @@
+  follow-up ASAP.
+* 100% durability of queries is permanently out of scope. We have to
+  accept the possibility of losing data under adverse conditions. The
+  reason for this design choice is that otherwise we would have to
+  persist the text of every statement (in either Cockroach or S3),
+  which we consider unacceptable for `SELECT` queries.
+
+## Overview
+
+The coordinator maintains a buffer of pending events related to
+statement execution. Whenever a prepared statement is created, a
+corresponding event is added to the buffer. Whenever a statement is
+executed, a system var called `STATEMENT_LOG_SAMPLING_RATE` is
+queried, and a random number between 0 and 1 is generated. If the
+random number is greater than the sampling rate, a description of the
+statement invocation is written into the buffer.
+
+Every so often (controlled by `STATEMENT_LOG_FLUSH_FREQUENCY`), the
+buffer is converted to updates that are written out to
+`mz_statement_execution_history` and `mz_prepared_statement_history`,
+with a column in the former referencing the latter. These system
+tables are persistent, but entries older than `STATEMENT_LOG_MAX_AGE`
+are periodically deleted.
+
+## Detailed description
+
+The coordinator maintains a buffer of pending events, which can be of
+the following variants:
+
+``` rust
+enum StatementLogEvent {
+    StatementPrepared {
+        id: Uuid,
+        sql: String,
+        name: Option<String>,
+    },
+    StatementExecutionBegan {
+        id: Uuid,
+        prepared_statement_id: Uuid,        
+        params: Vec<Datum>,
+        when: DateTime,
+        // Additional metadata might eventually
+        // go here, e.g. `plan: Option<OptimizedMirRelationExpr>`
+    },
+    StatementExecutionFinished {
+        id: Uuid,
+        when: DateTime,
+        n_results: Option<usize>,
+    },
+}
+```
+
+The `StatementPrepared` and `StatementExecutionBegan` events are
+generated and stored in the buffer in `fn handle_execute`, after
+checking a random value against `STATEMENT_LOG_SAMPLING_RATE` . This
+requires keeping track of the prepared statement bound to each name,
+and whether it has been logged yet (as we don't want to log each
+statement several times). The `StatementExecutionFinished` events are
+generated in several different places, depending on the type of
+statement. For example, the event for `SELECT` queries is generated in
+response to the `ControllerResponse::PeekResponse` message.
+
+The coordinator also holds on to an interval object whose tick rate is
+controlled by `STATEMENT_LOG_FLUSH_FREQUENCY`. It listens for ticks
+on that interval in its main `serve` loop; when such a tick occurs, it
+emits data to the system tables, as follows:
+
+* Each `StatementPrepared` event generates an entry in
+  `mz_prepared_statement_history`.
+* Each `StatementExecutionBegan` event generates an entry in
+  `mz_statement_execution_history`, and records that entry for
+  retraction as part of a later update.
```

Comment:
Later retraction? Just due to `statement_log_max_age`, or something else?

Reply:
It's because we need to update the row (to record how long the statement took) when `StatementExecutionEnded` comes in, and updating a row involves a retraction and insertion.